src/System.Runtime.Extensions/tests/System/Math.cs

```diff
@@ -1,134 +1,654 @@
+        /// <param name="actual">The value to be compared against</param>
+        /// <param name="allowedVariance">The total variance allowed between the expected and actual results.</param>
+        /// <exception cref="EqualException">Thrown when the values are not equal</exception>
+        public static void AssertEqual(double expected, double actual, double variance)
         {
-            Assert.Equal(0.54030230586814, Math.Cos(1.0), 10);
-            Assert.Equal(1.0, Math.Cos(0.0));
-            Assert.Equal(0.54030230586814, Math.Cos(-1.0), 10);
-            Assert.Equal(double.NaN, Math.Cos(double.NaN));
-            Assert.Equal(double.NaN, Math.Cos(double.PositiveInfinity));
-            Assert.Equal(double.NaN, Math.Cos(double.NegativeInfinity));
+            if (double.IsNaN(expected))
+            {
+                if (double.IsNaN(actual))
+                {
+                    return;
+                }
+
+                throw new EqualException(ToString(expected), ToString(actual));
+            }
+            else if (double.IsNaN(actual))
+            {
+                throw new EqualException(ToString(expected), ToString(actual));
+            }
+
+            if (double.IsNegativeInfinity(expected))
+            {
+                if (double.IsNegativeInfinity(actual))
+                {
+                    return;
+                }
+
+                throw new EqualException(ToString(expected), ToString(actual));
+            }
+            else if (double.IsNegativeInfinity(actual))
+            {
+                throw new EqualException(ToString(expected), ToString(actual));
+            }
+
+            if (double.IsPositiveInfinity(expected))
+            {
+                if (double.IsPositiveInfinity(actual))
+                {
+                    return;
+                }
+
+                throw new EqualException(ToString(expected), ToString(actual));
+            }
+            else if (double.IsPositiveInfinity(actual))
+            {
+                throw new EqualException(ToString(expected), ToString(actual));
+            }
+
+            if (IsNegativeZero(expected))
+            {
+                if (IsNegativeZero(actual))
+                {
+                    return;
+                }
+
+                if (IsPositiveZero(variance) || IsNegativeZero(variance))
+                {
+                    throw new EqualException(ToString(expected), ToString(actual));
+                }
+
+                // When the variance is not ±0.0, then we are handling a case where
+                // the actual result is expected to not be exactly -0.0 on some platforms
+                // and we should fallback to checking if it is within the allowed variance instead.
+            }
+            else if (IsNegativeZero(actual))
+            {
+                // Do nothing, since the actual result could still be within the allowed variance
```

Comment:
I should actually be checking `if (IsPositiveZero(variance) || IsNegativeZero(variance))` and throwing if true, as is done above. That way exactly precise values are handled.